Enabling Server-side Tokenization, Prompt Generation, and Streamlined Benchmarking

stisiTT opened the pull request:
This PR introduces a comprehensive set of improvements to the tt-inference-server's tokenization and prompt generation capabilities, enabling more robust and flexible benchmarking workflows with standardized configuration management.

## Key Features Added

### 🔧 **Server-side Tokenization Support**
- Added `entokenize()` and `detokenize()` methods to `PromptClient` for server-side tokenization
- Enables tokenization operations to be performed directly on the inference server rather than client-side
- Provides consistency between client and server token handling
- Supports both local and remote tokenization workflows

### 🎯 **CleanedPromptGenerator Class**
- New `CleanedPromptGenerator` class that creates stable token sequences through multiple encode/decode cycles
- Supports both client-side (local) and server-side tokenization modes
- Generates reproducible random prompts with configurable token lengths
- Implements fallback tokenizer support for improved reliability
- Provides batch generation capabilities for multiple prompt sequences

### 📊 **Streamlined Benchmarking Script**
- New `cleaned_benchmark_serving.py` that leverages the CleanedPromptGenerator
- Integrates server-side tokenization for more accurate benchmarking
- Provides cleaned, stable prompts that eliminate tokenization inconsistencies
- Includes comprehensive metrics collection and reporting
- Supports async request processing with configurable concurrency
- **Standardized EnvironmentConfig usage** aligned with workflow patterns

### 🛠 **Enhanced Utilities**
- `example_server_tokenization.py` demonstrates server-side tokenization usage
- Improved error handling and logging throughout the tokenization pipeline
- Better integration between prompt generation and client-server communication
- Support for various prompt distributions (fixed, uniform, normal)
- **Consistent configuration management** across all benchmark utilities

## Technical Improvements

- **Stability**: Multiple encode/decode cycles ensure token sequence stability
- **Flexibility**: Seamless switching between client-side and server-side tokenization
- **Reproducibility**: Configurable seeding for consistent prompt generation
- **Robustness**: Fallback mechanisms and comprehensive error handling
- **Performance**: Async processing and optimized request handling
- **Configuration Consistency**: Standardized EnvironmentConfig usage with environment variable fallbacks
- **Workflow Integration**: Aligned configuration patterns with existing benchmark scripts

## Use Cases

- **Load Testing**: Generate consistent prompts for server stress testing
- **Benchmarking**: Compare performance across different tokenization methods
- **Development**: Test server tokenization endpoints during development
- **Research**: Generate reproducible datasets for model evaluation

## Files Added/Modified

### New Files
- `utils/cleaned_benchmark_serving.py` - Streamlined benchmarking script with server-side tokenization
- `utils/cleaned_prompt_generator.py` - Class-based stable prompt generation utility
- `utils/example_server_tokenization.py` - Example demonstrating server-side tokenization usage

### Modified Files
- `utils/prompt_client.py` - Added `entokenize()` and `detokenize()` methods for server-side tokenization
- `utils/prompt_generation.py` - Enhanced with improved tokenization support
- `utils/README.md` - Updated documentation for new utilities
- `utils/cleaned_benchmark_serving.py` - Standardized EnvironmentConfig usage and added JWT secret support

## Implementation Details

### CleanedPromptGenerator
The `CleanedPromptGenerator` class implements a novel approach to generating stable token sequences:

1. **Random Token Generation**: Creates initial random token sequences
2. **Decode-Encode Cycles**: Performs multiple decode/encode cycles to stabilize tokens
3. **Dual Mode Support**: Works with both client-side and server-side tokenization
4. **Reproducible Results**: Uses configurable seeding for consistent outputs

### Server-side Tokenization
The new tokenization endpoints provide:

- **`entokenize(prompt)`**: Convert text to tokens using server tokenizer
- **`detokenize(tokens)`**: Convert tokens back to text using server tokenizer
- **Consistency**: Ensures client and server use identical tokenization
- **Error Handling**: Robust error handling with fallback mechanisms

### Benchmarking Improvements
The cleaned benchmarking script offers:

- **Stable Prompts**: Uses CleanedPromptGenerator for consistent test data
- **Server Integration**: Leverages server-side tokenization for accuracy
- **Async Processing**: Efficient concurrent request handling
- **Comprehensive Metrics**: Detailed performance analysis and reporting
- **Standardized Configuration**: Uses EnvironmentConfig with environment variable fallbacks
- **Workflow Consistency**: Follows same configuration pattern as `run_evals.py` and `run_benchmarks.py`
- **Flexible Authentication**: Supports JWT secret configuration via CLI arguments or environment variables

## Benefits

This enhancement significantly improves the server's testing and benchmarking capabilities while providing a foundation for more sophisticated prompt generation workflows. The server-side tokenization ensures perfect consistency between client and server token handling, eliminating a common source of discrepancies in benchmarking results.

The streamlined approach reduces complexity while increasing reliability, making it easier to conduct accurate performance evaluations and stress testing of the inference server. The standardized configuration management ensures consistent behavior across all benchmark utilities and seamless integration with existing workflow systems.

## Configuration Improvements

The latest updates standardize configuration management across the benchmarking ecosystem:

- **Environment Variable Support**: Automatic fallback to environment variables for all configuration options
- **CLI Override Capability**: Command-line arguments can override environment defaults
- **Workflow Integration**: Consistent with `run_evals.py` and `run_benchmarks.py` patterns
- **JWT Authentication**: Added `--jwt-secret` argument for flexible authentication configuration
- **Dynamic Configuration**: Runtime configuration updates based on command-line parameters

This ensures the cleaned benchmarking script integrates seamlessly with existing CI/CD pipelines and workflow automation systems.

stisiTT committed:
Cleaned Benchmarking Script
These files provide code to call to a server's tokenizer and pass back and forth to stabilize a randomly generated token.

stisiTT commented:
The following is from the new, "cleaned" benchmarks script

```
============ Serving Benchmark Result ============
Successful requests:                     1         
Benchmark duration (s):                  99.17     
Total input tokens:                      65280     
Total generated tokens:                  256       
Request throughput (req/s):              0.01      
Output token throughput (tok/s):         2.58      
Total Token throughput (tok/s):          660.82    
---------------------- TTFT ----------------------
Mean ttft (ms):                          45334.66  
Median ttft (ms):                        45334.66  
P99 ttft (ms):                           45334.66  
---------------------- TPOT ----------------------
Mean tpot (ms):                          211.12    
Median tpot (ms):                        211.12    
P99 tpot (ms):                           211.12    
---------------------- ITL -----------------------
Mean itl (ms):                           317.31    
Median itl (ms):                         227.46    
P99 itl (ms):                            434.09    
---------------------- E2EL ----------------------
Mean e2el (ms):                          99171.06  
Median e2el (ms):                        99171.06  
P99 e2el (ms):                           99171.06  
==================================================

============ Serving Benchmark Result ============
Successful requests:                     1         
Benchmark duration (s):                  8.93      
Total input tokens:                      8064      
Total generated tokens:                  128       
Request throughput (req/s):              0.11      
Output token throughput (tok/s):         14.33     
Total Token throughput (tok/s):          917.09    
---------------------- TTFT ----------------------
Mean ttft (ms):                          2169.35   
Median ttft (ms):                        2169.35   
P99 ttft (ms):                           2169.35   
---------------------- TPOT ----------------------
Mean tpot (ms):                          53.23     
Median tpot (ms):                        53.23     
P99 tpot (ms):                           53.23     
---------------------- ITL -----------------------
Mean itl (ms):                           107.30    
Median itl (ms):                         106.93    
P99 itl (ms):                            115.32    
---------------------- E2EL ----------------------
Mean e2el (ms):                          8930.19   
Median e2el (ms):                        8930.19   
P99 e2el (ms):                           8930.19   
==================================================

stisiTT committed:
Added capture_trace and vocab_size update

tstescoTT commented:
Can you integrate utils/cleaned_benchmark_serving.py as the script version used in https://github.com/tenstorrent/tt-inference-server/blob/dev/workflows/workflow_venvs.py#L219? I dont think we need to duplicate the script in this repo.

stisiTT committed:
feat: standardize EnvironmentConfig usage in cleaned_benchmark_serving.py

Align cleaned_benchmark_serving.py with workflow pattern for EnvironmentConfig

- Update setup_server_client() to accept args and model_config parameters
- Use EnvironmentConfig() default constructor with environment variable fallbacks
- Override config values dynamically from command-line arguments
- Add --jwt-secret argument to match other workflow scripts
- Follow same pattern as run_evals.py and run_benchmarks.py for consistency

This change enables better integration with the workflow system and provides
flexible configuration via environment variables or CLI arguments.

stisiTT commented on utils/prompt_configs.py:
```diff
@@ -46,7 +46,7 @@ def get_mesh_device():
 class EnvironmentConfig:
     vllm_model: str = os.environ.get(
-        "HF_MODEL_REPO_ID", "meta-llama/Llama-3.1-70B-Instruct"
+        "HF_MODEL_REPO_ID", "meta-llama/Llama-3.1-70B-Instruct" # TODO: This was causing an authorization bug if a workflow didn't previously set it
```
This was just during development and was part of my last remaining TODO items. This was since replaced (about 20 minutes ago) with a more appropriate use of EnvironmentConfig that's more in line with how run_evals.py and run_benchmarks.py uses it.

stisiTT commented:
It's a net-new benchmarking script altogether, or at least that's the hope. clean_prompt_generator.py and the updated prompt_client.py are technically the main changes. 

I'll make a separate PR for the changes to the vLLM fork.

FWIW:
2 would be patching the original download script, 3 would be redirecting to this repo for a download. 

1 is me changing the original but leaving download calls to it as-is.

tstescoTT commented:
Add vLLM benchmark script in branch off of https://github.com/tenstorrent/vllm/tree/skhorasgani/dev_rebase_upstream7

stisiTT commented:
Since we moved this to the vLLM fork, is there any use or purpose of these functions/processes? Do we want any of these in tt-inference-server?

tstescoTT commented:
> Since we moved this to the vLLM fork, is there any use or purpose of these functions/processes? Do we want any of these in tt-inference-server?

I think we can close this and maybe revisit adding the `investigations` directory. For now keeping this branch for reference is good. Feel free to close the PR if you agree.

stisiTT commented:
I agree. I'll maybe add this to the backlog if it's not already there, but we can get to it if/when we need to.
